File: src/android/com/adobe/phonegap/push/utils/CordovaCallbackContexts.java
package com.adobe.phonegap.push.utils;

import com.adobe.phonegap.push.PushPlugin;
import com.adobe.phonegap.push.logs.Logger;

import org.apache.cordova.CallbackContext;
import java.util.ArrayList;
import java.util.HashMap;

public class CordovaCallbackContexts {

    private static String TAG = PushPlugin.PREFIX_TAG + " (PushPlugin)";

    /**
     * Contains all the callback contexts for each Activity used by the application
     */
    private static HashMap<String, ArrayList<CallbackContext>> callbackContextsMap = new HashMap<>();

    public static ArrayList<CallbackContext> getCallbacksByActivity(String activityName) {
        create(activityName);
        return callbackContextsMap.get(activityName);
    }

    public static void create(String activityName) {
        if (!callbackContextsMap.containsKey(activityName)) {
            callbackContextsMap.put(activityName, new ArrayList<>());
        }
    }

    public static void remove(String activityName) {
        if (callbackContextsMap.containsKey(activityName)) {
            callbackContextsMap.remove(activityName);
        }
    }

    public static void add(CallbackContext callbackContext) {
        String activityName = PushPlugin.Companion.getCurrentActivityName();
        Logger.setDebug(true);
        ArrayList<CallbackContext> contexts = getCallbacksByActivity(activityName);
        int index = 0;
        int foundCallbackContext = -1;
        while (contexts.size() > index) {
            if (contexts.get(index).getCallbackId() == callbackContext.getCallbackId()) {
                foundCallbackContext = index;
                break;
            }
            index++;
        }
        if (foundCallbackContext == -1) {
            Logger.Debug(
                    TAG,
                    "addCallbackContent",
                    "Registering new CallbackContext #" + callbackContext.getCallbackId() + " inside Activity " + activityName
            );
            contexts.add(callbackContext);
        }
    }
}

File: src/android/com/adobe/phonegap/push/utils/CordovaCallbackContexts.kt
package com.adobe.phonegap.push.utils

import com.adobe.phonegap.push.PushPlugin.Companion.getCurrentActivityName
import com.adobe.phonegap.push.logs.Logger.setDebug
import com.adobe.phonegap.push.logs.Logger.Debug
import com.adobe.phonegap.push.PushPlugin
import org.apache.cordova.CallbackContext
import java.util.ArrayList
import java.util.HashMap

object CordovaCallbackContexts {
    private const val TAG = PushPlugin.PREFIX_TAG + " (PushPlugin)"

    /**
     * Contains all the callback contexts for each Activity used by the application
     */
    private val callbackContextsMap = HashMap<String, ArrayList<CallbackContext>>()
    fun getCallbacksByActivity(activityName: String): ArrayList<CallbackContext> {
        create(activityName)
        return callbackContextsMap[activityName]!!
    }

    fun create(activityName: String) {
        if (!callbackContextsMap.containsKey(activityName)) {
            callbackContextsMap[activityName] = ArrayList()
        }
    }

    fun remove(activityName: String) {
        if (callbackContextsMap.containsKey(activityName)) {
            callbackContextsMap.remove(activityName)
        }
    }

    fun add(callbackContext: CallbackContext) {
        val activityName = getCurrentActivityName()
        setDebug(true)
        val contexts = getCallbacksByActivity(activityName)
        var index = 0
        var foundCallbackContext = -1
        while (contexts.size > index) {
            if (contexts[index].callbackId === callbackContext.callbackId) {
                foundCallbackContext = index
                break
            }
            index++
        }
        if (foundCallbackContext == -1) {
            Debug(
                TAG,
                "addCallbackContent",
                "Registering new CallbackContext #" + callbackContext.callbackId + " inside Activity " + activityName
            )
            contexts.add(callbackContext)
        }
    }
}